### ISSUE TITLE
panicked at 'Value always exists' in zinc-compiler

### ISSUE THREAD

**Issue by oberstet** [matter-labs/zinc#12]
Not sure if the is the right repo or way/format to submit a bug report .. I'm not even 100% sure it is a bug, but I am running into a panic when building this code:

https://github.com/crossbario/zkmoloch/blob/eacd7c1e13c59e59c30bef366a457790262ac2ba/zk_moloch/src/main.zn#L75

on this machine/toolset:

```
(cpy39_1) oberstet@intel-nuci7:~/scm/crossbario/zkmoloch/zk_moloch$ zargo --version
zargo 0.2.3
(cpy39_1) oberstet@intel-nuci7:~/scm/crossbario/zkmoloch/zk_moloch$ uname -a
Linux intel-nuci7 5.4.0-77-generic #86-Ubuntu SMP Thu Jun 17 02:35:03 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
(cpy39_1) oberstet@intel-nuci7:~/scm/crossbario/zkmoloch/zk_moloch$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
```

Here is the traceback:

```
zargo build
   Compiling zkMoloch v0.1.0
thread '<unnamed>' panicked at 'Value always exists', zinc-compiler/src/semantic/element/value/mod.rs:665:22
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   2: core::option::expect_failed
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:1260:5
   3: zinc_compiler::semantic::element::value::Value::not
   4: zinc_compiler::semantic::element::Element::not
   5: zinc_compiler::semantic::analyzer::expression::Analyzer::unary
   6: zinc_compiler::semantic::analyzer::expression::Analyzer::traverse
   7: zinc_compiler::semantic::analyzer::expression::Analyzer::analyze
   8: zinc_compiler::semantic::analyzer::expression::list::Analyzer::analyze
   9: zinc_compiler::semantic::analyzer::expression::Analyzer::evaluate
  10: zinc_compiler::semantic::analyzer::expression::Analyzer::traverse
  11: zinc_compiler::semantic::analyzer::expression::Analyzer::right_local
  12: zinc_compiler::semantic::analyzer::expression::Analyzer::traverse
  13: zinc_compiler::semantic::analyzer::expression::Analyzer::analyze
  14: zinc_compiler::semantic::analyzer::expression::block::Analyzer::analyze
  15: zinc_compiler::semantic::analyzer::expression::conditional::Analyzer::analyze
  16: zinc_compiler::semantic::analyzer::expression::Analyzer::evaluate
  17: zinc_compiler::semantic::analyzer::expression::Analyzer::traverse
  18: zinc_compiler::semantic::analyzer::expression::Analyzer::analyze
  19: zinc_compiler::semantic::analyzer::expression::block::Analyzer::analyze
  20: zinc_compiler::semantic::analyzer::statement::for::Analyzer::define
  21: zinc_compiler::semantic::analyzer::expression::block::Analyzer::analyze
  22: zinc_compiler::semantic::analyzer::statement::fn::Analyzer::define
  23: zinc_compiler::semantic::scope::item::type::Type::define
  24: zinc_compiler::semantic::scope::item::Item::define
  25: zinc_compiler::semantic::scope::Scope::define
  26: zinc_compiler::semantic::analyzer::statement::contract::Analyzer::define
  27: zinc_compiler::semantic::scope::item::type::Type::define
  28: zinc_compiler::semantic::scope::item::Item::define
  29: zinc_compiler::semantic::scope::Scope::define
  30: zinc_compiler::semantic::analyzer::module::Analyzer::define
  31: zinc_compiler::semantic::scope::item::module::Module::define
  32: zinc_compiler::semantic::scope::item::Item::define
  33: zinc_compiler::semantic::analyzer::entry::Analyzer::define
  34: zinc_compiler::source::directory::Directory::compile
  35: zinc_compiler::bundler::Bundler::bundle
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'Thread synchronization is always successful: Any', zinc-compiler/src/znc/mod.rs:84:10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:1268:5
   3: znc::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[ERROR zargo] the subprocess failed with status exit code: 101                   
make: *** [Makefile:2: build] Fehler 1
```

Here is the full backtrace report https://gist.github.com/oberstet/e324c60f5c8f2b8f081672809635e3fa

**Comment by hedgar2017** [comment]
The Zinc 0.2 is discontinued, as well as the VM where the bug resides.

**Comment by oberstet** [comment]
Ok, thanks! Is there an update when 2.0 will be available? The last tag is 0.2.3 on this repo (which is what I used in above ..)

**Comment by hedgar2017** [comment]
0.2 will not, it is discontinued.
The next version is 0.3 Turing-complete :)

**Comment by oberstet** [comment]
^ thanks for the note rgd 0.3: Turing complete, awesome! Turing complete, EVM & Solidity compatible, ZK rollups: this is fantastic and unique!! can't wait;)

---

I'm though confused about version numbers

https://medium.com/matter-labs/zksync-2-0-roadmap-update-zkevm-testnet-in-may-mainnet-in-august-379c66995021

The "zksync 2.0" talked in this roadmap won't be tagged "2.0" in this repo as well?

Or, will it be, but 0.3 will be released before 2.0?

Essentially, I am looking for a clue when I can start with testing on sth public ..

**Comment by hedgar2017** [comment]
Nice to hear it :)

The Zinc release will be marked 0.3, and released not sooner than zkSync 2.0, since you'll need zkSync to deploy contracts.
